Local content item support

KludgeKML opened the pull request:
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

## What

Adds more structured support for loading content items from the local machine. By setting an environment variable and putting a content item in a particular directory, it will be loaded from that file instead of from a call to the content store.

## Why

There's currently a restricted version of this code working for Landing Pages (it only loads the blocks from the fixture, but includes fake values for the rest of the content item). This scaffolding has proved useful during Landing Pages development, but it's limited in scope and only works for Landing Pages. By replacing it with a more robust option, we can improve developer experience in the absence of a local version of content store, and make it easier to demo experimental content items on Heroku.

https://trello.com/c/FytCkByy/383-improve-contentitem-offline-loading, [Jira issue PNP-8577](https://gov-uk.atlassian.net/browse/PNP-8577)

## How

We add a new singleton class (ContentItemLoader) that is responsible for calls to GdsApi.content_store.get_content - this allows us to have a cache of items (simplifying the current Format Constraint code, which caches responses in the request env field), and to have a single point where those calls can be intercepted and swapped out for local code if the required environment variable is set. Setting ALLOW_LOCAL_CONTENT_ITEM_OVERRIDE to true will cause the loader to look in /config/local-content-items/path/to/the/item for any call to the content store (falling back to an actual call if the file isn't present).

## See Also:

* https://github.com/alphagov/govuk-docker/pull/793. (adds the env var to govuk-docker frontend apps)

KludgeKML committed:
Document Local Content Item support

KludgeKML committed:
Add ContentItemLoader class and tests

- ContentItemLoader replaces direct calls to GdsApi.content_store.content_item,
  (of which there are only 4 in the rest of the codebase), and provides two
  functions.
- First, it centralises caching of calls, so that that can be removed
  from the FormatRoutingConstraint classes and using request.env to save those
  values can be simplified.
- Second, if the ALLOW_LOCAL_CONTENT_ITEM_OVERRIDE env var is set true, it
  allows loading content items from a file in /lib/data/local-content-items,
  which gives developers an extra option for local development or preview apps
  when working with content types that have not crystalised yet or where
  publishing support is not yet present. It should not be used in production.
- The load method is slightly odd in that it will either return an API response
  or an exception (note: return the exception, not raise it!). This allows us
  to cache errors in a similar way to the way the routing constraints used to.
  It might be there are better ways to handle this, but for the moment this is
  a minimal change to maintain the current behaviour.

KludgeKML committed:
Update routing constraints to use ContentItemLoader

- Now that ContentItemLoader is handling the caching, we can remove that layer of code / responsibility from the constraints.
- ContentItemLoader either returns the adapter response or the error that it caught, so querying for those classes gives us a "was it or wasn't it an error" check.
- We also simplify the spec tests slightly to use more idiomatic RSpec.

KludgeKML committed:
Update controllers to use ContentItemLoader

- Now that ContentItemLoader handles caching we can use that rather than the stuff that the format constraints were putting into the request env.
- LandingPageController's scaffolding is still needed for the moment, but we can remove it soon (since part of this project is about replacing that scaffolding with a more generally useful one).
- Content items in controllers are now always loaded from `content_item_path`, which defaults to the request path, and is overridden if necessary for items with multiple paths on one content item.

leenagupte commented on docs/local-content-items.md:
```diff
@@ -0,0 +1,11 @@
+
+To support this, set:
+
+`ALLOW_LOCAL_CONTENT_ITEM_OVERRIDE=true`
```
It'd be good if there was an accompanying change in govuk-docker that sets this.

KludgeKML replied:
https://github.com/alphagov/govuk-docker/pull/793

KludgeKML replied:
Done (and added it to app.json)

leenagupte commented on app/controllers/landing_page_controller.rb:
```diff
@@ -5,8 +5,14 @@ class LandingPageController < ContentItemsController
+  # SCAFFOLDING: can be removed when basic content items are available
+  # from content-store
+  def old_scaffolding_content_item
+    ContentItemLoaderGdsApi.content_store.content_item(request.path).to_h
```
This doesn't look right. There isn't actually a class called `ContentItemLoaderGdsApi` is there?

The tests are passing, so either this is fine somehow, or this bit of code is not being tested.

KludgeKML replied:
Yeah, not sure what was going on here. I've fixed it now!

leenagupte reviewed (commented):
I like the idea in principle. I have a few inline comments that I think need to be fixed before this will work.

KludgeKML commented on lib/content_item_loader.rb:
```diff
@@ -0,0 +1,47 @@
+      File.exist?(local_json_filename(base_path)) || File.exist?(local_yaml_filename(base_path))
+    end
+
+    def local_json_filename(base_path)
+      Rails.root.join("#{LOCAL_ITEMS_PATH}#{base_path}.json")
+    end
+
+    def local_yaml_filename(base_path)
+      Rails.root.join("#{LOCAL_ITEMS_PATH}#{base_path}.yml")
+    end
```
We actually don't really need to do that - dashes are okay in file names, and it allows us to be simpler in matching the path. You can see an example here: https://github.com/alphagov/frontend/pull/4387/files

leenagupte commented on lib/content_item_loader.rb:
```diff
@@ -0,0 +1,47 @@
+    end
+
+    def local_file?(base_path)
+      File.exist?(local_json_filename(base_path)) || File.exist?(local_yaml_filename(base_path))
```
All of the methods apart from `load` should be made private as they're not being used or tested elsewhere.

leenagupte commented on app/controllers/landing_page_controller.rb:
```diff
@@ -5,9 +5,16 @@ class LandingPageController < ContentItemsController
+  # SCAFFOLDING: can be removed when basic content items are available
+  # from content-store
+  def old_scaffolding_content_item
+    result = ContentItemLoader.load(request.path)
```
It's great that you've fixed this. I think it would be good to add a test for this scaffolding too.

leenagupte replied:
They can't have been thorough enough because nothing caught the attempted call to `ContentItemLoaderGdsApi.content_store.content_item(request.path).to_h` which doesn't exist!

leenagupte reviewed (commented):
I had a bit more time to review, so had a closer look at the ContentItemLoader class and added a few inline suggestions for refactoring it.

There's also a request for some tests around the scaffolding if this is going to become a permanent feature.

There's nothing major though and the overall approach is great 🎉

KludgeKML committed:
Add ALLOW_LOCAL_CONTENT_ITEM_OVERRIDE=true to heroku config and startup.sh

KludgeKML committed:
Add a helper to all tests that resets the ContentItemLoader cache

- Because the cache is at class level, it's very aggressive and would otherwise interfere with what people would normally expect about tests (ie that in two unrelated tests you could use the same slug to point to different things). So we default here to just clearing the cache before each test.
- Do it manually in shared tests, which have their own setup that might conflict with "before" blocks in the system specs that call them.

leenagupte commented on spec/unit/content_item_loader_spec.rb:
```diff
@@ -0,0 +1,67 @@
+      context "with a local JSON file" do
+        let!(:item_request) { stub_content_store_has_item("/my-json-item") }
+
+        it "loads from that" do
```
Not a blocker, but this isn't a great test name! 

It should be probably be something like `it "loads content from the json file not content store". Or something.

KludgeKML replied:
Ta! Fixed.

leenagupte commented on spec/unit/content_item_loader_spec.rb:
```diff
@@ -0,0 +1,67 @@
+      context "with a local YAML file" do
+        let!(:item_request) { stub_content_store_has_item("/my-yaml-item") }
+
+        it "loads from that file" do
```
Ditto here about the test name, but again, not a blocker.

KludgeKML replied:
Ta! Fixed.

leenagupte reviewed (approved):
There are a couple of non-blocking comments about the test names, otherwise it looks good. 🎉
